// hypixel/client/skyblock/base.js
const https = require('https')
const fs = require('fs')
exports.collections = () => {
    return getFriends = new Promise((resolve, reject) => {
      var options = {
        method: 'GET',
        hostname: 'api.hypixel.net',
        path: `/resources/skyblock/collections`
      }
      let request = https.request(options, (res) => {
        let chunks = []
        res.on("data", chunk => chunks.push(chunk))
        res.on("end", chunk => {
          let collections = JSON.parse(Buffer.concat(chunks).toString())
          if (collections.success) {
            resolve(collections["collections"])
          } else {
            throw new Error('hi')
          }
        })
      })
      request.end()
    })
}

exports.skills = () => {
    return getSkills = new Promise((resolve, reject) => {
      var options = {
        method: 'GET',
        hostname: 'api.hypixel.net',
        path: `/resources/skyblock/skills`
      }
      let request = https.request(options, (res) => {
        let chunks = []
        res.on("data", chunk => chunks.push(chunk))
        res.on("end", chunk => {
          let skills = JSON.parse(Buffer.concat(chunks).toString())
          if (skills.success) {
            resolve(skills["skills"])
          } else {
            throw new Error('hi')
          }
        })
      })
      request.end()
    })
}

exports.news = (key) => {
  return getNews = new Promise((resolve, reject) => {
    var options = {
      method: 'GET',
      hostname: 'api.hypixel.net',
      path: `/skyblock/news`,
      headers: {
        'API-Key': key
      }
    }
    let request = https.request(options, (res) => {
      let chunks = []
      res.on("data", chunk => chunks.push(chunk))
      res.on("end", chunk => {
        let news = JSON.parse(Buffer.concat(chunks).toString())
        if (news.success) {
          resolve(news["items"])
        } else {
          throw new Error('hi')
        }
      })
    })
    request.end()
  })
}
